domain/storageprovisioning/service/volume.go

```diff
@@ -0,0 +1,233 @@
+// that have a very week data model about what a volume attachment is
+// attached to.
+//
+// All returned values will have either the machine name or unit name value
+// filled out in the [storageprovisioning.VolumeAttachmentID] struct.
+func (s *Service) GetVolumeAttachmentIDs(
+	ctx context.Context, uuids []string,
+) (map[string]storageprovisioning.VolumeAttachmentID, error) {
+	return s.st.GetVolumeAttachmentIDs(ctx, uuids)
+}
+
+// WatchModelProvisionedVolumes returns a watcher that emits volume IDs,
+// whenever a model provisioned volume's life changes.
+func (s *Service) WatchModelProvisionedVolumes(
+	ctx context.Context,
+) (watcher.StringsWatcher, error) {
+	ns, initialQuery := s.st.InitialWatchStatementModelProvisionedVolumes()
+	return s.watcherFactory.NewNamespaceWatcher(
+		initialQuery,
+		eventsource.NamespaceFilter(ns, changestream.All))
+}
+
+// WatchMachineProvisionedVolumes returns a watcher that emits volume IDs,
+// whenever a machine provisioned volume's life changes for the given machine.
+//
+// The following errors may be returned:
+// - [github.com/juju/juju/core/errors.NotValid] when the provided machine uuid
+// is not valid.
+// - [github.com/juju/juju/domain/machine/errors.MachineNotFound] when no
+// machine exists for the provided machine UUUID.
+func (s *Service) WatchMachineProvisionedVolumes(
+	ctx context.Context, machineUUID machine.UUID,
+) (watcher.StringsWatcher, error) {
+	if err := machineUUID.Validate(); err != nil {
+		return nil, errors.Capture(err)
+	}
+	netNodeUUID, err := s.st.GetMachineNetNodeUUID(ctx, machineUUID)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	lifeGetter := func(ctx context.Context) (map[string]life.Life, error) {
+		return s.st.GetVolumeLifeForNetNode(ctx, netNodeUUID)
+	}
+
+	ns, initialLifeQuery := s.st.InitialWatchStatementMachineProvisionedVolumes(netNodeUUID)
+	initialQuery, mapper := MakeEntityLifePrerequisites(initialLifeQuery, lifeGetter)
+	filter := eventsource.PredicateFilter(ns, changestream.All, eventsource.EqualsPredicate(netNodeUUID))
+
+	w, err := s.watcherFactory.NewNamespaceMapperWatcher(
+		initialQuery, mapper, filter)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return w, nil
+}
+
+// WatchModelProvisionedVolumeAttachments returns a watcher that emits volume
+// attachment UUIDs, whenever a model provisioned volume attachment's life
+// changes.
+func (s *Service) WatchModelProvisionedVolumeAttachments(
+	ctx context.Context,
+) (watcher.StringsWatcher, error) {
+	ns, initialQuery := s.st.InitialWatchStatementModelProvisionedVolumeAttachments()
+	return s.watcherFactory.NewNamespaceWatcher(initialQuery,
+		eventsource.NamespaceFilter(ns, changestream.All))
+}
+
+// WatchMachineProvisionedVolumeAttachments returns a watcher that emits volume
+// attachment UUIDs, whenever a machine provisioned volume attachment's life
+// changes for the given machine.
```

Comment:
... whenever the given machine's provisioned volume attachment's life changes.

Same elsewhere